src/hotspot/share/prims/jvmtiImpl.cpp

```diff
@@ -962,10 +962,12 @@ JvmtiDeferredEvent JvmtiDeferredEventQueue::dequeue() {
 }
 
 void JvmtiDeferredEventQueue::post(JvmtiEnv* env) {
-  // Post and destroy queue nodes
+  // Post events while nmethods are still in the queue and can't be unloaded or made zombie
+  for (QueueNode* node = _queue_head; node != NULL; node = node->next()) {
+    node->event().post_compiled_method_load_event(env);
+  }
```

Comment:
Can't you dequeue() immediately after calling post_compiled_method_load_event()?

Reply:
Seems that dequeue in for-loop deletes the node which posted. It is possible to update the loop to have update dequeue in the same iteration however, I don't think it is a good idea to mix iterator/deletion in the same loop. What is the reason for this change?

Reply:
Just to save iterating the queue twice. And that is what the original loop did - you just have to switch the order of processing and deleting.